Fix connection leak in isotope

howardjohn opened the pull request:
Fixes https://github.com/istio/istio/issues/16735

Basically if there was an error (5xx), then we never closed the request
body, leading to a leak of a connection.

I also had to update the dockerfile and add go.mod

istio-testing commented:
@howardjohn: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_tools | fed0616314a043e61d12e1de0c7659ce9bbf44c5 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_tools/851/lint_tools/1005) | `/test lint_tools`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@howardjohn: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_tools | fed0616314a043e61d12e1de0c7659ce9bbf44c5 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_tools/851/lint_tools/1005) | `/test lint_tools`
gencheck_tools | 67f35e8a922f8df0bd875ae1ffb3e4227d160f28 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_tools/851/gencheck_tools/757) | `/test gencheck_tools`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

carolynhu commented:
Why do you delete some of those licenses file? like fortio. Don't we need them?

carolynhu commented on perf/load/values.yaml:
```diff
@@ -1,4 +1,4 @@
-serviceGraphImage: tahler/isotope-service:1
+serviceGraphImage: gcr.io/istio-testing/isotope:0.0.1
```
Who is responsible for building and pushing isotope image? Also the versioning.

howardjohn replied:
We could set up the full CI/CD but this image has changed 1 time in the past 1.5years+, so I am not sure its needed?

howardjohn commented:
/hold

this is breaking something, requests are not getting sent properly for some reason. Will investigate and update tomorrow

howardjohn commented:
Found the issue, the probability code was wrong. I am not sure how it worked before, possibly when we updated deps it was broken? I also switched the logging to istio log to remove the dependency on fortio

howardjohn commented:
> Why do you delete some of those licenses file? like fortio. Don't we need them?

This is autogenerated based on what is used in the code. We no longer are importing fortio

sdake commented on isotope/service/Dockerfile:
```diff
@@ -1,20 +1,22 @@
-    go build -a -installsuffix cgo -o ./main ./main.go
+RUN GOARCH=amd64 CGO_ENABLED=0 GOOS=linux go build -o isotope_service ./service
+
+RUN ls -la
```
debug output

sdake commented on isotope/service/pkg/srv/executable.go:
```diff
@@ -25,13 +25,16 @@ import (
 )
 
+func init() {
+	rand.Seed(time.Now().UnixNano())
```
If you do a rev of this PR, you may consider adding a comment here that random number usage in this code is not for a cryptographic purpose. rand.Seed() (may or should) trigger some various security linters. These annoying linters don't add much value, but are required by various CISO (chief information security officer) peeps for no good reason. There may be multiple CISOs in one firm...

howardjohn commented:
Updated per comments, PTAL

howardjohn commented:
@sdake please take a look. Thanks!

howardjohn commented:
@sdake please take a look. Thanks!

mandarjog commented:
Thanks John this is awesome! This issue was open for such a long time !
